### PR TITLE
fix: Timeout unfulfilled request to decodingInfo and requestMediaKeySystemAccess

### LINKED ISSUE #7680: [Android][WebView] Calls to decodingInfo or requestMediaKeySystemAccess do not resolve or reject at least without RESOURCE_PROTECTED_MEDIA_ID
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?** Yes

**If the problem is related to FairPlay, have you read the tutorial?** N/A

**What version of Shaka Player are you using?** v4.12.2

**Can you reproduce the issue with our latest release version?** Yes

**Can you reproduce the issue with the latest code from `main`?** Yes

**Are you using the demo app or your own custom app?** Tried both, fails on both, fails even on https://shaka-player-demo.appspot.com/support.html

**If custom app, can you reproduce the issue using our demo app?** Yes

**What browser and OS are you using?** Telegram WebView on Samsung Galaxy S23 Ultra (Android 14)

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**
Mozilla/5.0 (Linux; Android 14; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.6723.107 Mobile Safari/537.36 Telegram-Android/11.4.2 (Samsung SM-S918B; Android 14; SDK 34; HIGH)

**What are the manifest and license server URIs?** N/A

**What configuration are you using?  What is the output of `player.getNonDefaultConfiguration()`?** Default

**What did you do?**

Try to use shaka on Telegram's WebView on Android (I think without [RESOURCE_PROTECTED_MEDIA_ID](https://developer.android.com/reference/android/webkit/PermissionRequest#RESOURCE_PROTECTED_MEDIA_ID)), or load https://shaka-player-demo.appspot.com/support.html
Even the DEMO https://shaka-player-demo.appspot.com/demo/ does not work in this environment
Any call to probeSupport gets hanged.

**What did you expect to happen?**
Calls to probeSupport not to get hanged on this environment even if the permission is not set.

**What actually happened?**

Calls to probeSupport get hanged.
Promise calls to decodingInfo or requestMediaKeySystemAccess do not resolve or reject on this environment.

**Are you planning send a PR to fix it?** Yes

### PR DESCRIPTION
**Pull request by CHaNGeTe**
On some (Android) WebView environments,
decodingInfo and requestMediaKeySystemAccess will
not resolve or reject, at least if RESOURCE_PROTECTED_MEDIA_ID is not set.
This is a workaround for that issue.

Closes #7680

### CODE REVIEW

File: lib/media/drm_engine.js [hunk 1]
```diff
@@ -1870,8 +1870,27 @@ shaka.media.DrmEngine = class {
...
+            navigator.requestMediaKeySystemAccess(keySystem, configs)
+                .then((access) => {
+                  resolved = true;
+                  resolve(access);
+                });
+            new shaka.util.Timer(() => {
+              if (!resolved) {
+                reject(new Error('Timeout requestMediaKeySystemAccess'));
+              }
+            }).tickAfter(TIMEOUT_FOR_CHECKACCESS_IN_SECONDS);
```
Comment (tykus160) [inline lib/media/drm_engine.js:1890]: You could add a utility method to simplify it, sth like:

```js
promiseWithTimeout(promise, timeout) {
  return Promise.race([
    promise,
    new Promise((_, reject) => new shaka.util.Timer(reject).tickAfter(timeout));
  ]);
}
```

**Review by tykus160 — CHANGES_REQUESTED** [review]

File: lib/media/drm_engine.js (hunk 1 above)
Comment (CHaNGeTe) [reply lib/media/drm_engine.js]: Added utility function promiseWithTimeout

**Comment by tykus160** [comment]
@CHaNGeTe could you sign CLA? Check https://github.com/shaka-project/shaka-player/pull/7682#issuecomment-2503775431

**Comment by CHaNGeTe** [comment]
> @CHaNGeTe could you sign CLA? Check [#7682 (comment)](https://github.com/shaka-project/shaka-player/pull/7682#issuecomment-2503775431)

I am working on it, asked the company for sign it

**Comment by tykus160** [comment]
@shaka-bot test

File: lib/media/drm_engine.js [hunk 2]
```diff
@@ -1907,13 +1915,22 @@ shaka.media.DrmEngine = class {
...
-
+        // On some (Android) WebView environments, decodingInfo will
+        // not resolve or reject, at least if RESOURCE_PROTECTED_MEDIA_ID
+        // is not set.  This is a workaround for that issue.
+        const TIMEOUT_FOR_DECODING_INFO_IN_SECONDS = 1;
         const decodingInfo =
-            await navigator.mediaCapabilities.decodingInfo(decodingConfig);
+          await shaka.util.Functional.promiseWithTimeout(
+              TIMEOUT_FOR_DECODING_INFO_IN_SECONDS,
+              navigator.mediaCapabilities.decodingInfo(decodingConfig),
```
Comment (tykus160) [inline lib/media/drm_engine.js:1925]: We're doing this call primarily in `StreamUtils`, I guess you should add it there as well.

File: lib/util/functional.js [hunk 3]
```diff
@@ -67,4 +69,29 @@ shaka.util.Functional = class {
...
+   * If |asyncProcess| takes too long, the returned Promise is rejected, but
+   * |asyncProcess| is still allowed to complete.
+   *
+   * @param {number} seconds
+   * @param {!Promise.<T>} asyncProcess
+   * @return {!Promise.<T>}
+   * @template T
+   */
+  static promiseWithTimeout(seconds, asyncProcess) {
+    return Promise.race([
```
Comment (tykus160) [inline lib/util/functional.js:90]: Can you add unit test around this method?

**Review by tykus160 — CHANGES_REQUESTED** [review]

File: lib/util/functional.js (hunk 3 above)
Comment (CHaNGeTe) [reply lib/util/functional.js:90]: Added Unit tests

**Comment by tykus160** [comment]
@shaka-bot test ce

**Review by tykus160 — APPROVED** [review]

**Review by avelad — APPROVED** [review]

**Pull request merged**